File: packages/compose/src/main/kotlin/ai/mobileagentos/compose/memory/InMemoryStore.kt
package ai.mobileagentos.compose.memory

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class InMemoryStore : MemoryStore {
    private val memories: MutableList<MemoryItem> = mutableListOf()
    private val mutex = Mutex()
    
    override suspend fun add(item: MemoryItem): Unit = mutex.withLock {
        memories.add(item)
    }
    
    override suspend fun retrieve(
        userId: String,
        query: String,
        limit: Int
    ): List<MemoryItem> = mutex.withLock {
        memories
            .filter { it.userId == userId }
            .filter { memory ->
                memory.content.contains(query, ignoreCase = true) ||
                memory.tags.any { tag -> tag.contains(query, ignoreCase = true) }
            }
            .sortedByDescending { it.importance }
            .take(limit)
    }
    
    override suspend fun list(userId: String): List<MemoryItem> = mutex.withLock {
        memories
            .filter { it.userId == userId }
            .sortedByDescending { it.timestamp }
    }
    
    override suspend fun clear(userId: String): Unit = mutex.withLock {
        memories.removeAll { it.userId == userId }
    }
}

File: packages/swift/Sources/MobileAgentOS/Memory/InMemoryStore.swift
import Foundation

/// Thread-safe in-memory storage for memories
public actor InMemoryStore: MemoryStore {
    private var memories: [MemoryItem] = []
    
    public init() {}
    
    public func add(_ item: MemoryItem) async {
        memories.append(item)
    }
    
    public func list() async -> [MemoryItem] {
        return memories
    }
    
    public func retrieve(query: String, limit: Int) async -> [MemoryItem] {
        let queryLowercased = query.lowercased()
        
        let scored = memories.map { memory -> (MemoryItem, Double) in
            var score = 0.0
            
            // Content match
            if memory.content.lowercased().contains(queryLowercased) {
                score += 0.5
            }
            
            // Tag match
            for tag in memory.tags {
                if queryLowercased.contains(tag.lowercased()) {
                    score += 0.3
                }
            }
            
            // Importance
            score += memory.importance * 0.2
            
            return (memory, score)
        }
        
        let sorted = scored
            .filter { $0.1 > 0 }
            .sorted { $0.1 > $1.1 }
            .prefix(limit)
            .map { $0.0 }
        
        return Array(sorted)
    }
    
    public func clear() async {
        memories.removeAll()
    }
}